### ISSUE TITLE
<T extends Partial<Post>> does not produce compiler error when partial object with non-exist property is given

### ISSUE THREAD

**Issue by pleerock** [Microsoft/TypeScript#14287]
**TypeScript Version:**  2.2.1

**Code**

```ts
class Post {
    id: number;
    title: string;
}

async function save1<T extends Partial<Post>>(entity: T): Promise<T>{
    return entity;
}

async function save2(entity: Partial<Post>): Promise<Partial<Post>>{
    return entity;
}

save1({ id: 1 }); // fine
save1({ title: "hello" }); // fine
save1({ a: 0, title: "hello" }); // this should give compiler error which is not expected
save2({ a: 0, title: "hello" }); // produces compile error which is expected
```

**Expected behavior:**

`save1({ a: 0, title: "hello" });` expression should produce compiler error

**Actual behavior:**

`save1({ a: 0, title: "hello" });` compiles fine

**Comment by mhegazy** [comment]
This is working as intended. in `save1` you are asking the compiler to infer a `T` from usage, then compare the result to a constraint which is `Partial<Post>`; and that is what the compiler does with `{ a: 0, title: "hello" }`.

I think what you are looking for is:
```ts
async function save3<T extends Post>(entity: Partial<T>): Promise<Partial<T>>{
    return entity;
}
```

**Comment by dkirchhof** [comment]
Hello, I have a similiar issue with a generic partial type:  
So in the example above i just want to return a type with all of the attributes i provide in the entity param, but save2 and save3 returns the whole T type with optional parameters.  
If i stay with the first save function - regardless of non existent properties - there is no autocompletion  while typing the object literal.   
  
In short: 
```
save1({ id: 1, | });
               ^
``` 
vscode or whatever ide should suggest me all post attributes.
